pkg/ccl/changefeedccl/sink_webhook_v2.go

```diff
@@ -178,20 +194,30 @@ func makeWebhookClient(
 }
 
 func (sc *webhookSinkClient) makePayloadForBytes(body []byte) (SinkPayload, error) {
-	req, err := http.NewRequestWithContext(sc.ctx, http.MethodPost, sc.url.String(), bytes.NewReader(body))
+	var finalBytes []byte
+	if sc.compression.enabled() {
+		var buf bytes.Buffer
+		codec, err := newCompressionCodec(sc.compression, &sc.settings.SV, &buf)
```

Comment:
Can you make the compression writers created once here in the sink client and Reset() and reused  in every request? Rather than allocating a new one for each request. This will necessitate changing the api in compression.go a bit.

From my reading of the docs I think this is the proper use of these apis but let me know if you disagree.

Reply:
Are you talking about coder/decoder pool? So whenever the coder is needed the client can request the coder from the pool and receive either new one if the pool is drained or existing if there is something in the pool. We'll also need to implement wrapper struct for codecs in order to Reset a buffer when the codec is received from the pool. Please confirm that I've got you correctly?

Reply:
I don't think we'll need a pool, just one encoder & decoder per webhookSinkClient, right?

Reply:
In that case, you might face unpredictable behaviour during the concurrent operations (prove me if I'm wrong because I might not clearly understand the use case). In general you either have to create a new instance each time it is needed or use sync.Pool that handles concurrent access

Reply:
Ah yes, you're right. We would need a sync.Pool in that case.

Reply:
The other problem is that you'll need multiple pools for each coder type (e.g. one for gzip coders, another for gzip decoders, and so on). I'm going to push the pool implementation later today. I belive that after this change compression.go should be converted into standalone package

Reply:
I've added coders pool. But I'm not sure where is the right place to init it. As far as I can see it should be initialised before the sinks. I'd appreciate any advice about that

Reply:
Thanks @massimo-ua. I think the api you've come up with is a little too easy to misuse, I think something like this might be better: https://goplay.tools/snippet/cUQP5XNq4g_k . What do you think?

Reply:
I have similar idea first but then I've checked the cloudstorage sink and found out that there is a File abstraction which in turn has codec property and in some cases in calls f.code.Close() and I didn't want to change the API significantly so I decided to override Close method so the clients don't even know that they must return the codec in the pool. But in general I'm ok with your suggestion since mine looks much more complicated and a bit overengineered

Reply:
I see. Yeah the reason I don't love that is because on the underlying writer/encoders, Close() does not denote the end of the life of the writer, since you can call Reset (which it occurs to me i did not do in my above example...). So there's room for confusion if the abstraction's Close method has different semantics from that.

Reply:
Ok. I see your point

Reply:
Potentially the compression API could have been much cleaner like this
```go
encode(algo compressionAlgo, dest io.Writer, sv *settings.Values) error
decode(algo compressionAlgo, src io.Reader) error
```
In that case there is no confusion because its not io.Reader/Writer but custom API
Current API is too open IMHO, and repetitive codec.Close(), done()

Reply:
Pool [implemented](https://github.com/cockroachdb/cockroach/pull/138872/commits/4acd79c49c044a98954fda18c2140a51a5613be3)

Reply:
I agree that the API could be cleaner. I had a look through the codebase and found this usage of a writer pool that I think maybe we could mimic here: https://github.com/asg0451/cockroach/blob/05c04054ed6ccda0d075f10ff973bf6f0af64951/pkg/rpc/snappy.go#L69-L108

This version does hide the fact that writer.Close does not mean what Close usually means, which on reflection maybe is a good thing.

What do you think? Sorry for the back and forth :/

Reply:
Ok. We will wrap coders/decoders and hide all the pull management under the hood of io.WriteCloser and ReadCloser but the API will remain the same, correct?
```go
codec, err := newCompressionCodec(sc.compression, &sc.settings.SV, &buf)
defer codec.Close()

reader, err := newDecompressionReader(algo, res.Body)
derer reader.Close()
```

Reply:
Yep, that's right. Thanks!